incubation-openmetrics.md (#615)
cncf/toc: incubation-openmetrics.md

# OpenMetrics proposal for CNCF incubation

## Background

> OpenMetrics specifies the de-facto standard for transmitting cloud-native metrics at scale, with support for both text representation and Protocol Buffers. It’s the wire format and data model Prometheus, Cortex, and Thanos rely on and are migrating more componentes to.

Reviewer: ```suggestion
OpenMetrics specifies the de-facto standard for transmitting cloud-native metrics at scale, with support for both text representation and Protocol Buffers. It’s the wire format and data model that Prometheus, Cortex, Thanos and many other open source projects rely on and further more are actively migrating to.
```

> At the same time, OpenMetrics is pursuing an RFC under IETF OPSAWG https://datatracker.ietf.org/doc/draft-richih-opsawg-openmetrics/ . 

Reviewer: ```suggestion
At the same time, OpenMetrics is pursuing an [https://datatracker.ietf.org/doc/draft-richih-opsawg-openmetrics/](RFC under IETF OPSAWG). 
```

Author: That's the wrong way around, will update PR with Markdown link.

OpenMetrics was created outside of Prometheus at the request of Dan Kohn and Chris Aniszczyk in 2015 and named OpenMetrics in 2016.

OpenMetrics came out 5th place in the CNCF End User Survey on Observability tools.

OpenMetrics entered the CNCF Sandbox in July 2018.

## Progress since joining the CNCF Sandbox

OpenMetrics carefully evolved and tightened Prometheus exposition format 0.0.4.

Here are some highlights:

 * Prometheus has supported and used OpenMetrics on the wire since 2018
 * Cortex and Thanos support it
 * DataDog has reimplemented OpenMetrics on their own without help from OpenMetrics
 * OpenMetrics has been submitted to IETF OPSAWG

>  * IETF OPSAWG showed very strong interest in quickly releasing OpenMetrics after a presentation at IETF 110

Reviewer: nit: For consistency add full stops to the end of each of these to match the last bullet point?

Author: Will do.

>  * Exemplar support has been merged into Prometheus main and will be part of the next release. This would not ever have happened without OpenMetrics. This directly enables users to correlate metrics with traces as cloud-native scale.

Reviewer: ```suggestion
 * Exemplar support has been merged into Prometheus main and will be part of the next release. This would not ever have happened without OpenMetrics. This directly enables users to correlate metrics with traces at cloud-native scale.
```

### Transition from Prometheus exposition format 0.0.4 to OpenMetrics 1.0

This is handled transparently by the Prometheus client libraries.

There are two changes, one breaking:

>  * Timestamps are in seconds, not microseconds. This was the only place where Prometheus used microseconds on the wire. As timestamps are not recommended in Prometheus expositions anyway and largely considered a hack/workaround for very few and specific situations, users are very unlikely to hit this.

Reviewer 2: s/micro/milli/g

Also, they'd have to expose the wrong content type for this to be an issue.

Author: Good points, will edit.

 * Counters must end in _total. Before, this was only strongly recommended. Client libraries handle this transparently.

### Feature roadmap for 2.x

> One highlight feature will be high-resolution histograms in close coordination with Prometheus. Also see https://docs.google.com/document/d/1cLNv3aufPZb3fNfaJgdaRBZsInZKKIHo9E6HinJVbpM/edit

Reviewer: ```suggestion
One highlight feature will be high-resolution histograms in close coordination with Prometheus. For more see the [design document](https://docs.google.com/document/d/1cLNv3aufPZb3fNfaJgdaRBZsInZKKIHo9E6HinJVbpM/edit).
```

Author: See above.

> ## Requirements for incubation

Reviewer: nit:
```suggestion
## Incubation 
```

> Please see DD by SIG O11y https://docs.google.com/document/d/1BN6ABvfbAMhJthH6xNOY5Lmls3jeE3kFbMgF04BzEKE

Reviewer: ```suggestion
Please see [due diligence by SIG O11y](https://docs.google.com/document/d/1BN6ABvfbAMhJthH6xNOY5Lmls3jeE3kFbMgF04BzEKE).
```
